packages/token/src/erc20/extensions/erc4626/erc4626.cairo

```diff
@@ -0,0 +1,604 @@
+    /// those `assets` to `receiver`.
+    #[derive(Drop, PartialEq, starknet::Event)]
+    pub struct Withdraw {
+        #[key]
+        pub sender: ContractAddress,
+        #[key]
+        pub receiver: ContractAddress,
+        #[key]
+        pub owner: ContractAddress,
+        pub assets: u256,
+        pub shares: u256
+    }
+
+    pub mod Errors {
+        pub const EXCEEDED_MAX_DEPOSIT: felt252 = 'ERC4626: exceeds max deposit';
+        pub const EXCEEDED_MAX_MINT: felt252 = 'ERC4626: exceeds max mint';
+        pub const EXCEEDED_MAX_WITHDRAW: felt252 = 'ERC4626: exceeds max withdraw';
+        pub const EXCEEDED_MAX_REDEEM: felt252 = 'ERC4626: exceeds max redeem';
+        pub const TOKEN_TRANSFER_FAILED: felt252 = 'ERC4626: token transfer failed';
+        pub const INVALID_ASSET_ADDRESS: felt252 = 'ERC4626: asset address set to 0';
+        pub const DECIMALS_OVERFLOW: felt252 = 'ERC4626: decimals overflow';
+    }
+
+    /// Constants expected to be defined at the contract level which configure virtual
+    /// assets and shares.
+    ///
+    /// `UNDERLYING_DECIMALS` should match the underlying asset's decimals. The default
+    /// value is `18`.
+    ///
+    /// `DECIMALS_OFFSET` corresponds to the representational offset between `UNDERLYING_DECIMALS`
+    /// and the vault decimals. The greater the offset, the more expensive it is for attackers to
+    /// execute an inflation attack.
+    ///
+    /// Requirements:
+    ///
+    /// - `UNDERLYING_DECIMALS`+ `DECIMALS_OFFSET` cannot exceed 255 (max u8).
+    ///
+    pub trait ImmutableConfig {
+        const UNDERLYING_DECIMALS: u8;
+        const DECIMALS_OFFSET: u8;
+
+        fn validate() {
+            assert(
+                Bounded::MAX - Self::UNDERLYING_DECIMALS >= Self::DECIMALS_OFFSET,
+                Errors::DECIMALS_OVERFLOW
+            )
+        }
+    }
+
+    /// Adjustments for fees expected to be defined on the contract level.
+    /// Defaults to no entry or exit fees.
+    /// To transfer fees, this trait needs to be coordinated with `ERC4626Component::ERC4626Hooks`.
+    pub trait FeeConfigTrait<TContractState> {
+        fn adjust_deposit(self: @ComponentState<TContractState>, assets: u256) -> u256 {
+            assets
+        }
+
+        fn adjust_mint(self: @ComponentState<TContractState>, assets: u256) -> u256 {
+            assets
+        }
+
+        fn adjust_withdraw(self: @ComponentState<TContractState>, assets: u256) -> u256 {
+            assets
+        }
+
+        fn adjust_redeem(self: @ComponentState<TContractState>, assets: u256) -> u256 {
+            assets
+        }
+    }
+
+    /// Sets custom limits to the target exchange type and is expected to be defined at the contract
+    /// level.
+    pub trait LimitConfigTrait<TContractState> {
+        fn deposit_limit(
+            self: @ComponentState<TContractState>, receiver: ContractAddress
+        ) -> Option::<u256> {
```

Comment:
We can omit `::` in such places, but it's also okay to leave it
```suggestion
        ) -> Option<u256> {
```

Reply:
Yes, indeed. Force of habit